docs/taskman-design.md

```diff
@@ -0,0 +1,379 @@
+
+```
+Public Key Algorithm: rsaEncryption
+Public Key Size: 4096 bit
+Signature Algorithm: sha256WithRSAEncryption
+```
+
+In a production environment, we would want these certificates to be generated, short-lived, and renewed through automation securely.
+
+#### Authorization
+Each certificate will be generated with a Common Name (CN) that will identify the user. The certificate's common name will be extracted by the gRPC server and associated with each created task. When a user makes a request against a task their common name will be extracted from the provided certificate on the request and must match the name associated with the task on the server in order for requests to be accepted.
+
+There will be an admin certificate that the gRPC server allows access to all tasks and operations on the server.
+
+### Privacy
+The gRPC server will only store data in memory for the life of the server and will be lost on server shutdown or restart. Any other privacy considerations are beyond the scope of this exercise.
+
+### Proto Specification
+This exercise will have a new protobuf representing the TaskManager service that exposes functionality to start, stop, get, and stream tasks.
+
+```package task_manager;
+
+// TaskManager is a service that exposes APIs to manage linux processes (tasks)
+service TaskManager {
+    // StartTask takes a task and arguments and starts a new task
+    rpc StartTask (StartTaskRequest) returns (StartTaskResponse);
+    // StopTask stops a running task by task ID
+    rpc StopTask (StopTaskRequest) returns (StopTaskResponse);
+    // GetTaskStatus gets the status of a task by task ID
+    rpc GetTaskStatus (TaskStatusRequest) returns (TaskStatusResponse);
+    // StreamTaskOutput streams the output of a task by task ID
+    rpc StreamTaskOutput (StreamTaskOutputRequest) returns (stream StreamTaskOutputResponse);
+}
+
+// JobStatus tracks status of job
+enum JobStatus {
+    // job is in an unknown error state
+    JOB_STATUS_UNKNOWN = 0;
+    // job is currently running
+    JOB_STATUS_STARTED = 1;
+    // job was manually stopped by client
+    JOB_STATUS_STOPPED = 2;
+    // job completed and exited normally
+    JOB_STATUS_EXITED_OK = 3;
+    // job exited with a non-zero status
+    JOB_STATUS_EXITED_ERROR = 4;
+}
+
+// StartTaskRequest contains the command and arguments to start a new task
+message StartTaskRequest {
+    // must be full path to the task to run e.g. "/bin/ls"
+    string task_path = 1;
+    // arguments to pass to the task e.g. ["-l", "-a"]
+    repeated string args = 2;
+}
+
+message StartTaskResponse {
+    // UUID v4 ID of the task generated by the server
+    string task_id = 1;
+}
+
+message StopTaskRequest {
+    // UUID v4 ID of the task generated by the server
+    string task_id = 1;
+}
+
+// StopTaskResponse returns true if the stop task request was successful; false otherwise
+message StopTaskResponse {
+    bool success = 1;
```

Comment:
Do you think it might be simpler to return a gRPC error on failure?

Reply:
Yes that seems like a better approach that follows better gRPC design